consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/ProposalValidator.java

```diff
@@ -0,0 +1,288 @@
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.core.Hash;
+
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class ProposalValidator {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private static final String ERROR_PREFIX = "Invalid Proposal Payload";
+
+  private final BlockValidator blockValidator;
+  private final ProtocolContext protocolContext;
+  private final int quorumMessageCount;
+  private final Collection<Address> validators;
+  private final ConsensusRoundIdentifier roundIdentifier;
+  private final Address expectedProposer;
+
+  public ProposalValidator(
+      final BlockValidator blockValidator,
+      final ProtocolContext protocolContext,
+      final int quorumMessageCount,
+      final Collection<Address> validators,
+      final ConsensusRoundIdentifier roundIdentifier,
+      final Address expectedProposer) {
+    this.blockValidator = blockValidator;
+    this.protocolContext = protocolContext;
+    this.quorumMessageCount = quorumMessageCount;
+    this.validators = validators;
+    this.roundIdentifier = roundIdentifier;
+    this.expectedProposer = expectedProposer;
+  }
+
+  public boolean validate(final Proposal msg) {
+
+    final ProposalPayloadValidator payloadValidator =
+        new ProposalPayloadValidator(
+            expectedProposer, roundIdentifier, blockValidator, protocolContext);
+
+    if (!payloadValidator.validate(msg.getSignedPayload())) {
+      LOG.info("{}: invalid proposal payload in proposal message", ERROR_PREFIX);
+      return false;
+    }
+
+    if (!validateProposalAndRoundChangeAreConsistent(msg)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  private boolean validateProposalAndRoundChangeAreConsistent(final Proposal proposal) {
+    final ConsensusRoundIdentifier proposalRoundIdentifier = proposal.getRoundIdentifier();
+
+    if (proposalRoundIdentifier.getRoundNumber() == 0) {
+      if (!validateRoundZeroProposalHasNoRoundChangesOrPrepares(proposal)) {
+        return false;
+      }
+
+      return validateBlockCoinbaseMatchesMsgAuthor(proposal);
+    } else {
+
+      if (!validateRoundChanges(proposal, proposal.getRoundChanges())) {
+        LOG.info("{}: failed to validate piggy-backed round change payloads", ERROR_PREFIX);
+        return false;
+      }
+
+      // The RoundChangePayloadValidator ensures the PreparedRound is less than targetRound
+      // therefore, no need to validate that here.
+      final Optional<SignedData<RoundChangePayload>> roundChangeWithLatestPreparedRound =
+          getRoundChangeWithLatestPreparedRound(proposal.getRoundChanges());
+
+      if (roundChangeWithLatestPreparedRound.isPresent()) {
+        final PreparedRoundMetadata metadata =
```

Comment:
There could be two RoundChange messages with the higher preparedRound but different preparedBlockHash.
In this case, as long as we have a valid set of prepare for one of them the message is valid.

Reply:
I'd argue that the Proposer should not have included both RoundChange messages in its justification, if the proposer is able to determine that a roundchange is invalid, then it should be used as justification to trigger a new round.